File: LeetCode/Array/JS/Leetcode556-next-greater-element-iii.js
/**
 * @param {number} n
 * @return {number}
 */
var nextGreaterElement = function (n) {
    let s = String(n).split('');
    let i = s.length - 1;
    while (i && s[i - 1] >= s[i]) i--;
    if (!i) return -1;
    let j = i;
    while (j + 1 < s.length && s[j + 1] > s[i - 1]) j++;
    [s[i - 1], s[j]] = [s[j], s[i - 1]]
    let reverse = (s, st) => {
        for (let i = st, j = s.length - 1; i < j; i++) {
            [s[i], s[j]] = [s[j], s[i]];
            j--;
        }
    }
    reverse(s, i)
    if (s.join('') > 2147483647) {
        return -1;
    }
    return s.join('');
};

File: LeetCode/Array/Python/Leetcode556-next-greater-element-iii.py
class Solution:
    def nextGreaterElement(self, n: int) -> int:
        s = list(str(n))
        i = len(s) - 1
        while i != 0 and s[i - 1] >= s[i]:
            i -= 1
        if i == 0: return -1
        j = i
        while j + 1 < len(s) and s[j + 1] > s[i - 1]:
            j += 1
        s[j], s[i - 1] = s[i - 1], s[j]
        # print(s[i:],s[-1:i - 1:-1])
        s[i:] = s[-1:i - 1:-1]
        # print(1 << 30)
        if int(''.join(s)) > ((1 << 31) - 1): return -1
        return int(''.join(s))
